### PR TITLE
Check an app's origin first for its runtime

### PR DESCRIPTION
**Pull request by mwleeds**
Currently when searching for a remote to provide the runtime for an app,
we search remotes in priority order. This commit makes it so we search
the remote providing the app first, then search the rest in priority
order. This behavior just seems logical, but the impetus for this change
was also to keep a unit test passing in eos-updater[1] after a port to
FlatpakTransaction.

In practice it's usually true that only one remote provides a runtime
and priorities aren't set at all, so this is an edge case that probably
doesn't come up much.

A unit test is included.

[1] https://github.com/endlessm/eos-updater/blob/eede0a8b9cb7c9cbb8539cfcaad9ce18913dcdd8/tests/test-update-install-flatpaks.c#L1919

### CODE REVIEW

**Comment by mwleeds** [comment]
I changed this so that we ignore noenumerate and disabled remotes even for the source remote that's being prioritized, which matches the existing behavior before this patch. And I added a commit to hopefully make the documentation around the no-enumerate option for remotes for clear.

**Comment by alexlarsson** [comment]
This seems like a potentially risky change. You can use it to install a rouge runtime with an id that matches one on flathub, even if flathub has higher prio, thus affecting other apps.

**Comment by mwleeds** [comment]
> This seems like a potentially risky change. You can use it to install a rouge runtime with an id that matches one on flathub, even if flathub has higher prio, thus affecting other apps.

That's true, but if you don't trust a remote to provide dependencies, shouldn't you set `nodeps` on it? Otherwise the situation you describe can happen anyway if the rogue remote was added before flathub and flathub doesn't have a priority set.

It's funny, [the commit message](https://github.com/endlessm/eos-updater/commit/2d8aada3dd34d5554baaa92b2df7817d0fc5c867) adding the unit test which spawned this patch also mentioned the goal of disempowering rogue remotes:

>      1. Dependency ref actions can come from any remote that is
>         currently configured. We should give more priority to the
>         remote that the source ref action came from, such that we
>         avoid accidentally installing an out of date or malicious
>         dependency from some other repo.
(For context ref actions are update/install/uninstall operations on flatpaks that happen as part of an OS update, or as a result of system configuration.)

**Comment by alexlarsson** [comment]
> That's true, but if you don't trust a remote to provide dependencies, shouldn't you set `nodeps` on it? Otherwise the situation you describe can happen anyway if the rogue remote was added before flathub and flathub doesn't have a priority set.

Sure, but that is a lot more work (and easier to miss) than setting in a single place a higher permission on the flathub remote (which may even be done by the distro).

Can we maybe search in priority order, but use the origin remote first in that priority level?

**Comment by mwleeds** [comment]
> > That's true, but if you don't trust a remote to provide dependencies, shouldn't you set `nodeps` on it? Otherwise the situation you describe can happen anyway if the rogue remote was added before flathub and flathub doesn't have a priority set.
> 
> Sure, but that is a lot more work (and easier to miss) than setting in a single place a higher permission on the flathub remote (which may even be done by the distro).
> 
> Can we maybe search in priority order, but use the origin remote first in that priority level?

Yeah, that seems reasonable. Relatedly, I checked to make sure the priority can't be set by a remote in `flatpak_dir_update_remote_configuration_for_state()` or in a `.flatpakrepo` file, because if it could be I think we'd want to warn the user.

**Comment by mwleeds** [comment]
Pushed a commit to migrate the tests from `g_assert` to `g_assert_true` to avoid `G_DISABLE_ASSERT`

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -12703,20 +12703,60 @@ flatpak_dir_list_enumerated_remotes (FlatpakDir   *self,
...
 char **
 flatpak_dir_search_for_dependency (FlatpakDir   *self,
+                                   const char   *prioritized_remote,
                                    const char   *runtime_ref,
                                    GCancellable *cancellable,
                                    GError      **error)
 {
   g_autoptr(GPtrArray) found = g_ptr_array_new_with_free_func (g_free);
   g_auto(GStrv) remotes = NULL;
+  g_autofree RemoteSortData *rsd = g_new0 (RemoteSortData, 1);
```
Comment (alexlarsson) [inline common/flatpak-dir.c:12747]: It feels unnecessary to allocate this. Just put it on the stack.

Comment (alexlarsson) [reply common/flatpak-dir.c:12747]: i.e.:
```
  RemoteSortData rsd = { NULL };
```

File: common/flatpak-dir.c [hunk 2]
```diff
@@ -12703,20 +12703,60 @@ flatpak_dir_list_enumerated_remotes (FlatpakDir   *self,
...
+   * putting a particular remote at the top of its priority level */
+  if (prio_a == prio_b)
+    {
+      if (strcmp (a_name, rsd->prioritized_remote) == 0)
+        return -1;
+      if (strcmp (b_name, rsd->prioritized_remote) == 0)
+        return 1;
+    }
+
+  return 0;
```
Comment (alexlarsson) [inline common/flatpak-dir.c:12735]: This isn't really a valid comparison function, as it doesn't define an order.

For example, if you had `a 0, b 1, c 1, d 2`.
Then a and d compare as equal, but they really need to have items between them.

So, even though we expect the list in prio order we need to compare the priorities anyway.

However for two items of the same prio we can return 0, because g_qsort is guaranteed to be stable (these days).

Comment (mwleeds) [reply common/flatpak-dir.c:12735]: Good point
